### PR TITLE
Menu.dialog

### PR DESCRIPTION
**Pull request by macoca**
This PR supports `dialog` on `Nri.Ui.Menu.V3`.

Required for https://linear.app/noredink/issue/GRO-538/homepage-login-div-should-be-a-dialog

It's derived from `disclosure` but kept both for compatibility and proper deprecation in the future.

Also, some minor cleanups were marked by elm-review while I looked at the modal.

https://user-images.githubusercontent.com/957103/200679365-77d2bc8a-9f86-4707-9c7e-4a3737410585.mov

### CODE REVIEW

File: tests/Spec/Nri/Ui/Menu.elm [hunk 1]
```diff
@@ -72,6 +72,35 @@ spec =
...
+                        |> pressTabKey { targetId = Just "last-button" }
+                        |> ensureViewHas (menuContentSelector menuContent)
+                        |> ProgramTest.done
+            , test "Selects lastId after back tab on firstId" <|
+                \() ->
+                    program [ Menu.dialog { firstId = "hello-button", lastId = "last-button" } ]
+                        |> clickMenuButton
+                        |> ensureViewHas (menuContentSelector menuContent)
+                        -- NOTE: unable to simulate pressTabKey with other targetId since those decoders will fail
+                        |> pressTabBackKey { targetId = Just "hellow-button" }
```
Comment (tesk9) [inline tests/Spec/Nri/Ui/Menu.elm:100]: Should "hellow-button" be "hello-button"?

File: src/Nri/Ui/Menu/V3.elm [hunk 2]
```diff
@@ -456,47 +478,77 @@ viewCustom config =
...
-                                            { isOpen = False
-                                            , focus = Nothing
-                                            }
-                                    , tabForwardAction =
-                                        config.focusAndToggle
-                                            { isOpen = False
-                                            , focus = Nothing
-                                            }
+                    Dialog { firstId, lastId } ->
+                        Key.onKeyDownPreventDefault
```
[the remark this answers is not in the document]
Comment (tesk9) [reply src/Nri/Ui/Menu/V3.elm:529]: Checked on the styleguide, and it works! My concern was unfounded 😄

File: tests/Spec/Nri/Ui/Menu.elm (hunk 1 above)
Comment (tesk9) [reply tests/Spec/Nri/Ui/Menu.elm:100]: I'm seeing test failures on this branch, is all

**Review by tesk9 — COMMENTED** [review]
Thank you for working on this! It's appreciated! And it's looking great 😄 

One small thing: I think when the dialog opens, we should move focus to the first focusable element

File: src/Nri/Ui/Menu/V3.elm (hunk 2 above)
[the remark this answers is not in the document]
Comment (macoca) [reply src/Nri/Ui/Menu/V3.elm:529]: No problem! I had to explicitly do that in other for the tab to properly wrap around. It was working fine for the back tab, but the forward tab wouldn't wrap without it.

File: src/Nri/Ui/Menu/V3.elm [hunk 3]
```diff
@@ -122,6 +122,13 @@ type alias ButtonConfig =
 type Purpose
     = NavMenu
     | Disclosure { lastId : String }
+    | Dialog ExitFocusManager
+
+
+type alias ExitFocusManager =
```
[the remark this answers is not in the document]
Comment (macoca) [reply src/Nri/Ui/Menu/V3.elm]: Thank you for the perfectly clear explanation!

**Comment by macoca** [comment]
I updated the video with the new behavior

File: src/Nri/Ui/Menu/V3.elm [hunk 4]
```diff
@@ -582,10 +616,18 @@ viewCustom config =
                                 { preventDefault = True
                                 , stopPropagation = True
                                 , message =
-                                    config.focusAndToggle
-                                        { isOpen = not config.isOpen
-                                        , focus = Nothing
-                                        }
+                                    case ( config.isOpen, config.purpose ) of
```
Comment (tesk9) [inline src/Nri/Ui/Menu/V3.elm:619]: Nice! Makes sense 👍

File: src/Nri/Ui/Menu/V3.elm [hunk 5]
```diff
@@ -599,6 +641,21 @@ viewCustom config =
                       else
                         AttributesExtra.none
                     ]
+                        ++ (case config.purpose of
```
Comment (tesk9) [inline src/Nri/Ui/Menu/V3.elm:644]: This case is nice & clear! Looks good

File: src/Nri/Ui/Modal/V11.elm [hunk 6]
```diff
@@ -516,11 +516,9 @@ view config attrsList model =
                 ]
                 |> List.singleton
                 |> Root.div
-                    (List.concat
-                        [ [ FocusTrap.toAttribute config.focusTrap ]
-                        , [ Attrs.css [ Css.position Css.relative, Css.zIndex (Css.int 100) ] ]
-                        ]
-                    )
+                    [ FocusTrap.toAttribute config.focusTrap
```
Comment (tesk9) [inline src/Nri/Ui/Modal/V11.elm:519]: thanks for cleaning up the lists here! much better

**Review by tesk9 — APPROVED** [review]
Fabulous! Thank you for all the great work on this! 🙏

**Pull request merged**